src/Utils.cpp

```diff
@@ -1569,3 +1569,165 @@ void Utils::xor_encdec(u_char *data, int data_len, u_char *key) {
+
+static uint16_t in_cksum(uint16_t *addr, unsigned len){
+  uint16_t answer = 0; 
+  /*
+   * Our algorithm is simple, using a 32 bit accumulator (sum), we add
+   * sequential 16 bit words to it, and at the end, fold back all the
+   * carry bits from the top 16 bits into the lower 16 bits.
+   */
+  uint32_t sum = 0; 
+  while (len > 1)  {
+    sum += *addr++;
+    len -= 2;
+  }
+
+  // mop up an odd byte, if necessary
+  if (len == 1) { 
+    *(unsigned char *)&answer = *(unsigned char *)addr ;
+    sum += answer;
+  }
+
+  // add back carry outs from top 16 bits to low 16 bits
+  sum = (sum >> 16) + (sum & 0xffff); // add high 16 to low 16
+  sum += (sum >> 16); // add carry
+  answer = ~sum; // truncate to 16 bits
+  return answer;
+}
+
+int Utils::ping(string target){
+	int s, i, cc, packlen, datalen = DEFDATALEN;
+	struct hostent *hp;
+	struct sockaddr_in to, from;
+	//struct protoent   *proto;
+	struct ip *ip;
+	u_char *packet, outpack[MAXPACKET];
+	char hnamebuf[MAXHOSTNAMELEN];
+	string hostname;
+	struct icmp *icp;
+	int ret, fromlen, hlen;
+	fd_set rfds;
+	struct timeval tv;
+	int retval;
+	struct timeval start, end;
+	int /*start_t, */end_t;
+	bool cont = true;
+
+	// TODO: handle IPv6
+
+	to.sin_family = AF_INET;
+
+	// try to convert as dotted decimal address, else if that fails assume it's a hostname
+	to.sin_addr.s_addr = inet_addr(target.c_str());
+	if(to.sin_addr.s_addr != (u_int)-1){
+		hostname = target;
+	}
+	else{
+		hp = gethostbyname(target.c_str());
+		if(!hp){
+			ntop->getTrace()->traceEvent(TRACE_ERROR, "unknown host %s\n", target.c_str());
+			return -1;
+		}
+		to.sin_family = hp->h_addrtype;
+		bcopy(hp->h_addr, (caddr_t)&to.sin_addr, hp->h_length);
+		strncpy(hnamebuf, hp->h_name, sizeof(hnamebuf) - 1);
+		hostname = hnamebuf;
+	}
+	packlen = datalen + MAXIPLEN + MAXICMPLEN;
+	if((packet = (u_char *)malloc((u_int)packlen)) == NULL){
+		ntop->getTrace()->traceEvent(TRACE_ERROR, "malloc error\n");
+		return -1;
+	}
+
+	if((s = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP)) < 0){
+		perror("socket");   /* probably not running as superuser */
+		return -1;
+	}
+
+	icp = (struct icmp *)outpack;
+	icp->icmp_type = ICMP_ECHO;
+	icp->icmp_code = 0;
+	icp->icmp_cksum = 0;
+	icp->icmp_seq = 12345;  /* seq and id must be reflected */
+	icp->icmp_id = getpid();
+
+
+	cc = datalen + ICMP_MINLEN;
+	icp->icmp_cksum = in_cksum((unsigned short *)icp,cc);
+
+	gettimeofday(&start, NULL);
+
+	i = sendto(s, (char *)outpack, cc, 0, (struct sockaddr*)&to, (socklen_t)sizeof(struct sockaddr_in));
+	if(i < 0 || i != cc){
+		if(i < 0){
+			perror("sendto error");
+		}
+		ntop->getTrace()->traceEvent(TRACE_INFO, "wrote %s %d chars, ret= %d\n", hostname.c_str(), cc, i);
+	}
+
+	// Watch stdin (fd 0) to see when it has input.
+	FD_ZERO(&rfds);
+	FD_SET(s, &rfds);
+	// Wait up to one seconds.
+	tv.tv_sec = 1;
+	tv.tv_usec = 0;
+
+	while(cont){
+		retval = select(s+1, &rfds, NULL, NULL, &tv);
```

Comment:
are you sure you need the select? you don't have to monitor more than one file descriptor. This can be overkilling.